File: practice-compare-robyn/robyn-api/models/message.py
from sqlalchemy import Uuid, Column, String, Integer, ForeignKey, UniqueConstraint
from sqlalchemy.orm import relationship
from config.database import Base
from uuid import uuid4

from models.time_mixin import TimeMixin


class MessagePremission(Base):
    __tablename__ = "message_premission"

    id = Column(Uuid(as_uuid=True), primary_key=True, default=uuid4)
    level = Column(Integer, unique=True)
    description = Column(String, default="")

    def __repr__(self):
        return f"<MessagePremission(id={self.id}, level={self.level}, description={self.description})>"

    @staticmethod
    def base_premission():
        return [
            MessagePremission(level=0, description="Public"),
            MessagePremission(level=1, description="Protected"),
            MessagePremission(level=2, description="Private"),
        ]


class Message(Base, TimeMixin):
    __tablename__ = "messages"

    id = Column(Uuid(as_uuid=True), primary_key=True, default=uuid4)
    author_id = Column(Uuid(as_uuid=True), ForeignKey("users.id"))
    title = Column(String)
    content = Column(String)
    permission_level = Column(Integer, ForeignKey("message_premission.level"))
    history = relationship("ReadHistory", back_populates="message", cascade="all, delete-orphan")

    def __repr__(self):
        return (
            f"<Message(id={self.id}, author_id={self.author_id}, "
            f"title={self.title}, content={self.content}, "
            f"permission_level={self.permission_level})>"
        )


class ReadHistory(Base, TimeMixin):
    __tablename__ = "read_history"

    id = Column(Uuid(as_uuid=True), primary_key=True, default=uuid4)
    user_id = Column(Uuid(as_uuid=True), ForeignKey("users.id"))
    message_id = Column(Uuid(as_uuid=True), ForeignKey("messages.id", ondelete="CASCADE"))
    __table_args__ = (UniqueConstraint("user_id", "message_id", name="uq_user_message"),)
    message = relationship("Message", back_populates="history")

    def __repr__(self):
        return f"<ReadHistory(id={self.id}, user_id={self.user_id}, message_id={self.message_id})>"

File: practice-compare-robyn/gin-api/models/message.go
package models

import (
	"compare-with-robyn/config"

	"github.com/google/uuid"
)

type MessagePremission struct {
	ID          uuid.UUID `gorm:"type:uuid;primaryKey;default:uuid_generate_v4()"`
	Level       uint8     `gorm:"not null;unique"`
	Description string    `gorm:"default:''"`
}

type Message struct {
	ID                uuid.UUID `gorm:"type:uuid;primaryKey;default:uuid_generate_v4()"`
	AuthorID          uuid.UUID `gorm:"type:uuid;not null"`
	Title             string
	Content           string
	PermissionLevel   uint8             `gorm:"not null" json:"permission_level"`
	User              User              `gorm:"foreignKey:AuthorID;constraint:OnDelete:SET NULL"`
	MessagePremission MessagePremission `gorm:"foreignKey:PermissionLevel;references:Level"`
	MixinTime
}

type ReadHistory struct {
	ID        uuid.UUID `gorm:"type:uuid;primaryKey;default:uuid_generate_v4()"`
	UserID    uuid.UUID `gorm:"type:uuid;not null;uniqueIndex:idx_user_message"`
	MessageID uuid.UUID `gorm:"type:uuid;not null;uniqueIndex:idx_user_message"`
	User      User      `gorm:"foreignKey:UserID"`
	Message   Message   `gorm:"foreignKey:MessageID;constraint:OnDelete:CASCADE"`
	MixinTime
}

func BasePremission() []MessagePremission {
	return []MessagePremission{
		{Level: 1, Description: "Public"},
		{Level: 2, Description: "Protected"},
		{Level: 3, Description: "Private"},
	}
}

func (m *Message) Create() error {
	return config.DB.Create(m).Error
}

func (m *Message) Delete() error {
	return config.DB.Where("author_id = ?", m.AuthorID).Delete(m).Error
}
